src/script.c

```diff
@@ -0,0 +1,422 @@
+        } else {
+            *err = sdsnew("Script attempted to access a non local key in a "
+                    "cluster node");
+        }
+        return C_ERR;
+    }
+    return C_OK;
+}
+
+static void scriptEmitMultiIfNeeded(scriptRunCtx *r_ctx) {
+    /* If we are using single commands replication, we need to wrap what
+     * we propagate into a MULTI/EXEC block, so that it will be atomic like
+     * a Lua script in the context of AOF and slaves. */
+    client *c = r_ctx->c;
+    if (!(r_ctx->flags & SCRIPT_EVAL_REPLICATION)
+         && !(r_ctx->flags & SCRIPT_MULTI_EMMITED)
+         && !(r_ctx->original_client->flags & CLIENT_MULTI)
+         && (r_ctx->flags & SCRIPT_WRITE_DIRTY)
+         && ((r_ctx->repl_flags & PROPAGATE_AOF)
+             || (r_ctx->repl_flags & PROPAGATE_REPL)))
+    {
+        execCommandPropagateMulti(r_ctx->original_client->db->id);
+        r_ctx->flags |= SCRIPT_MULTI_EMMITED;
+        /* Now we are in the MULTI context, the lua_client should be
+         * flag as CLIENT_MULTI. */
+        c->flags |= CLIENT_MULTI;
+    }
+}
+
+/* set RESP for a given r_ctx */
+int scriptSetResp(scriptRunCtx *r_ctx, int resp) {
+    if (resp != 2 && resp != 3) {
+        return C_ERR;
+    }
+
+    r_ctx->c->resp = resp;
+    return C_OK;
+}
+
+/* set Repl for a given r_ctx
+ * either: PROPAGATE_AOF | PROPAGATE_REPL*/
+int scriptSetRepl(scriptRunCtx *r_ctx, int repl) {
+    if ((repl & ~(PROPAGATE_AOF | PROPAGATE_REPL)) != 0) {
+        return C_ERR;
+    }
+    r_ctx->repl_flags = repl;
+    return C_OK;
+}
+
+/* Call a Redis command.
+ * The reply is written to the r_ctx client and it is
+ * up to the engine to take and parse.
+ * The err out variable is set only if error occurs and describe the error.
+ * If err is set on reply is written to the r_ctx client. */
+void scriptCall(scriptRunCtx *r_ctx, robj* *argv, int argc, sds *err) {
+    client *c = r_ctx->c;
+
+    /* Setup our fake client for command execution */
+    c->argv = argv;
+    c->argc = argc;
+    c->user = r_ctx->original_client->user;
+
+    /* Process module hooks */
+    moduleCallCommandFilters(c);
+    argv = c->argv;
+    argc = c->argc;
+
+    struct redisCommand *cmd = lookupCommand(argv, argc);
+    if (scriptVerifyCommandArity(cmd, argc, err) != C_OK) {
+        return;
+    }
+
+    c->cmd = c->lastcmd = cmd;
+
+    /* There are commands that are not allowed inside scripts. */
+    if (!server.script_disable_deny_script && (cmd->flags & CMD_NOSCRIPT)) {
+        *err = sdsnew("This Redis command is not allowed from script");
+        return;
+    }
+
+    if (scriptVerifyACL(c, err) != C_OK) {
+        return;
+    }
+
+    if (scriptVerifyWriteCommandAllow(r_ctx, err) != C_OK) {
+        return;
+    }
+
+    if (scriptVerifyOOM(r_ctx, err) != C_OK) {
+        return;
+    }
+
+    if (cmd->flags & CMD_WRITE) {
+        /* signify that we already change the data in this execution */
+        r_ctx->flags |= SCRIPT_WRITE_DIRTY;
```

Comment:
It looks like we do this today, but just because a command is a write, doesn't mean it dirties the dataset. I suppose this means we might send a multi-exec filled with commands that don't execute. Not a correctness problem I presume, just weird.

Reply:
I am not sure what you mean here, do you refer to this issue: https://github.com/redis/redis/issues/8478?

Reply:
No, if we send a write command we mark the script as dirty. We don't need to though, we only need to care if the write command actually does something to the dataset. I'll open an separate issue for this, it's not related but something I noticed.

Reply:
I see, so I guess I can resolve this one?

Reply:
yes, this is not related to here, let's add it to #8478